File: 2024/Day5/2.py
from collections import defaultdict, deque

def parse_rules(rule_lines):
    rules = {}
    for rule in rule_lines:
        x, y = map(int, rule.split("|"))
        rules[(x, y)] = True  # X must come before Y
    return rules

def is_valid_update(update, rules):
    for i in range(len(update)):
        for j in range(i + 1, len(update)):
            x, y = update[i], update[j]
            if (y, x) in rules:  # If there's a rule Y|X, this order is invalid
                return False
    return True

def find_middle_page(update):
    return update[len(update) // 2]

def reorder_update(update, rules):
    """Reorder the pages in the update according to the rules."""
    # Create a directed graph from the rules
    graph = defaultdict(list)
    in_degree = defaultdict(int)
    pages_in_update = set(update)

    for x, y in rules:
        if x in pages_in_update and y in pages_in_update:
            graph[x].append(y)
            in_degree[y] += 1
            if x not in in_degree:
                in_degree[x] = 0

    # Topological sort
    queue = deque([node for node in update if in_degree[node] == 0])
    sorted_order = []

    while queue:
        node = queue.popleft()
        sorted_order.append(node)
        for neighbor in graph[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    return sorted_order

def main():
    with open("input.txt", "r") as file:
        lines = file.readlines()

    # Separate rules and updates
    rules_section = []
    updates_section = []
    parsing_rules = True
    for line in lines:
        line = line.strip()
        if not line:
            parsing_rules = False
            continue
        if parsing_rules:
            rules_section.append(line)
        else:
            updates_section.append(line)

    # Parse rules
    rules = parse_rules(rules_section)

    # Process updates
    invalid_updates = []
    for update_line in updates_section:
        update = list(map(int, update_line.split(",")))
        if not is_valid_update(update, rules):
            invalid_updates.append(update)

    total_middle_sum = 0
    for update in invalid_updates:
        corrected_update = reorder_update(update, rules)
        total_middle_sum += find_middle_page(corrected_update)
    print(f"Total Middle Sum: {total_middle_sum}")

if __name__ == "__main__":
    main()

File: 2024/Day5/2.swift
import Foundation

let filePath = "input.txt"

struct Rule: Hashable {
    let x: Int
    let y: Int
}

func parseRules(_ ruleLines: [String]) -> Set<Rule> {
    var rules = Set<Rule>()
    for rule in ruleLines {
        let parts = rule.split(separator: "|").compactMap { Int($0) }
        if parts.count == 2 {
            rules.insert(Rule(x: parts[0], y: parts[1]))
        }
    }
    return rules
}

func isValidUpdate(_ update: [Int], rules: Set<Rule>) -> Bool {
    for i in 0..<update.count {
        for j in (i + 1)..<update.count {
            let x = update[i], y = update[j]
            if rules.contains(Rule(x: y, y: x)) { // If there's a rule Y|X, this order is invalid
                return false
            }
        }
    }
    return true
}

func findMiddlePage(_ update: [Int]) -> Int {
    return update[update.count / 2]
}

func reorderUpdate(_ update: [Int], rules: Set<Rule>) -> [Int] {
    var graph = [Int: [Int]]()
    var inDegree = [Int: Int]()
    let pagesInUpdate = Set(update)

    // Build the graph and compute in-degrees
    for rule in rules {
        if pagesInUpdate.contains(rule.x) && pagesInUpdate.contains(rule.y) {
            graph[rule.x, default: []].append(rule.y)
            inDegree[rule.y, default: 0] += 1
            if inDegree[rule.x] == nil {
                inDegree[rule.x] = 0
            }
        }
    }

    // Perform topological sort
    var queue = [Int]()
    for page in update {
        if inDegree[page, default: 0] == 0 {
            queue.append(page)
        }
    }

    var sortedOrder = [Int]()
    while !queue.isEmpty {
        let node = queue.removeFirst()
        sortedOrder.append(node)
        for neighbor in graph[node, default: []] {
            inDegree[neighbor, default: 0] -= 1
            if inDegree[neighbor, default: 0] == 0 {
                queue.append(neighbor)
            }
        }
    }

    return sortedOrder
}

do {
  let content = try String(contentsOfFile: filePath, encoding: .utf8)

  let lines = content.split(separator: "\n").map { String($0) }

  var rulesSection = [String]()
  var updatesSection = [String]()
  var parsingRules = true

  for line in lines {
    if line.contains(",") {
      parsingRules = false
    }
    if parsingRules {
      rulesSection.append(line)
    } else {
      updatesSection.append(line)
    }
  }

  // Parse rules
  let rules = parseRules(rulesSection)

  var invalidUpdates = [[Int]]()
  for updateLine in updatesSection {
        let update = updateLine.split(separator: ",").compactMap { Int($0) }
        if !isValidUpdate(update, rules: rules) {
            invalidUpdates.append(update)
        }
  }

  var totalMiddleSum = 0
    for update in invalidUpdates {
        let correctedUpdate = reorderUpdate(update, rules: rules)
        totalMiddleSum += findMiddlePage(correctedUpdate)
    }

  print("Total Middle Sum: \(totalMiddleSum)")

} catch {
  print("Error: \(error)")
}
